### ISSUE TITLE
_mm_loadu_si128 parameter has wrong type

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#21044]
|  |  |
| --- | --- |
| Bugzilla Link | [20670](https://llvm.org/bz20670) |
| Version | trunk |
| OS | FreeBSD |
| Reporter | LLVM Bugzilla Contributor |
| CC | @topperc,@RKSimon,@noloader,@rnk |
| Fixed by commit(s) | r353802 |

## Extended Description 
In <emmintrin.h>, _mm_loadu_si128 is declared as:
```
static __inline__ __m128i __attribute__((__always_inline__, __nodebug__)) _mm_loadu_si128(__m128i const *__p)
```
i.e., taking a (__m128i const *) parameter.  This is a bug, since __m128i is 16-byte aligned, and the whole point of the _mm_loadu_si128 is that it performs *unaligned* loads -- so you're forcing people to write code which produces "cast increases required alignment" warnings.

I realize that this was Intel's bug originally, but I don't see any reason why you can't fix it.

**Comment by RKSimon** [comment]
> > Worth noting that it's not just _mm_loadu_si128; _mm_lddqu_si128
> > _mm_storeu_si128, _mm256_loadu_si128, and _mm256_lddqu_si256 have the same
> > problem.
> > Looks like Intel figured it out for AVX-512 and just used void const*.
> 
> It looks like we still need to use the __m128i_u / __m256i_u types on the lddqu intrinsics.

Candidate Patch: https://reviews.llvm.org/D121815

**Comment by RKSimon** [comment]
lddqu intrinsics were fixed in 4e4f839ac2a9002df504f1bed8383c2d7d72e86e
